src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborWriterTests.Array.cs

```diff
@@ -186,5 +186,23 @@ public static void WriteEndArray_AfterStartMap_ShouldThrowInvalidOperationExcept
             writer.WriteStartMap(definiteLength: 0);
             Assert.Throws<InvalidOperationException>(() => writer.WriteEndArray());
         }
+
+        [Theory]
+        [InlineData(CborConformanceLevel.Rfc7049Canonical)]
+        [InlineData(CborConformanceLevel.Ctap2Canonical)]
+        internal static void WriteArray_IndefiniteLength_UnsupportedConformanceLevel_ShouldThrowInvalidOperationException(CborConformanceLevel level)
+        {
+            using var writer = new CborWriter(level);
+            Assert.Throws<InvalidOperationException>(() => writer.WriteStartArrayIndefiniteLength());
```

Comment:
Would it make sense to allow this function to work, then the equivalent end method backfills the length?

Sure, it may require data shifting, but it lets a caller drain an unknown length IEnumerable without double-enumerating.

Reply:
It should be relatively straightforward for maps and arrays, but would require much more book-keeping in the case of indefinite-length strings. I'd be in favor of supporting the former case only (AFAICT there we haven't had a strong use case for supporting indefinite-length strings so far).

Should we support this for canonical conformance levels only, or would it make sense to make it the out-of-the-box behaviour everywhere?

Reply:
I can get behind only supporting it for maps/arrays at first.  Maybe "Indefinite" can mean indefinite and "Unknown" can be the feature I'm asking for?

```C#
writer.WriteStartArrayUnknownLength();
...
writer.WriteEndArray();
```

(Or, perhaps nicer)

```C#
using (writer.WriteStartArrayUnknownLength())
{
    ...
}
```

Reply:
Assuming we do this for every conformance level, it might be the case that we could just call it `WriteStartArray()`, since the end result would be indistinguishable from a definite-length write. It would also offer symmetry vis-a-vis the map reader api.

Reply:
I don't see the "we'll count it for you" WriteStartArray in this PR currently.  Is it expected in this one, or a followup?

Reply:
It's been pushed to this PR just now.